host/frontend/host-orchestrator/jsonws.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+//     Unless required by applicable law or agreed to in writing, software
+//     distributed under the License is distributed on an "AS IS" BASIS,
+//     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//     See the License for the specific language governing permissions and
+//     limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+// A websocket connection that can send and receive json objects.
+// Using websockets requires some boilerplate code, even using gorilla, which is
+// encapsulated behind this object.
+// All methods in this struct are thread safe.
+type JsonWs struct {
+	send chan []byte
+	recv chan []byte
+	cl   chan bool
+	conn *websocket.Conn
+}
+
+func NewJsonWs(w http.ResponseWriter, r *http.Request) *JsonWs {
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Println(err)
+		return nil
+	}
+	send, recv, cl := make(chan []byte, 100), make(chan []byte, 100), make(chan bool)
+	go writePump(conn, send)
+	go readPump(conn, recv, cl)
+	return &JsonWs{send: send, recv: recv, cl: cl, conn: conn}
+}
+
+func (ws *JsonWs) Send(val interface{}) bool {
+	select {
+	case _, _ = <-ws.cl:
+		// A message on this channel means the ws is closed
```

Comment:
nit: This isn't ever reading a message, but is instead getting the "closed indicator":

https://stackoverflow.com/a/16105721

Reply:
Reworded